### PR TITLE
fix: use relative redirects

### LINKED ISSUE #33: Missing etag downloading `TheBloke/Llama-2-7b-Chat-GGUF` + `llama-2-7b-chat.Q4_K_M.gguf` (Tokio)
Got a `Err(MissingHeader("etag"))` when downloading `TheBloke/Llama-2-7b-Chat-GGUF` + `llama-2-7b-chat.Q4_K_M.gguf`.

Following test reproduces it on main.

```rust
    #[tokio::test]
    async fn reproduce_missing_etag() {
        let tmp = TempDir::new();
        let api = ApiBuilder::new()
            .with_token(None)
            .with_progress(false)
            .with_cache_dir(tmp.path.clone())
            .build()
            .unwrap();
        let result = api
            .model("TheBloke/Llama-2-7b-Chat-GGUF".to_string())
            .download("llama-2-7b-chat.Q4_K_M.gguf")
            .await;
        assert_ok!(result);
    }
```

> `thread 'api::tokio::tests::reproduce_missing_etag' panicked at 'assertion failed: Err(MissingHeader("etag"))', src/api/tokio.rs:828:9`

More directly hitting the `metadata` function:

```rust
    #[tokio::test]
    async fn metadata() {
        let tmp = TempDir::new();
        let api = ApiBuilder::new()
            .with_token(None)
            .with_progress(false)
            .with_cache_dir(tmp.path.clone())
            .build()
            .unwrap();

        let repo = api.model("TheBloke/Llama-2-7b-Chat-GGUF".to_string());
        let file_url = &repo.url("llama-2-7b-chat.Q4_K_M.gguf");

        println!("{file_url}");

        let result = api
            .metadata(file_url)
            .await;

        assert_ok!(result);
    }
```

also fails - with the url `https://huggingface.co/TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf`

which when `curl`ed is a `302`, which does not get followed as metadata used `no_redirect_client`.

swapping out `no_redirect_client` for `client` in `metadata` yeilds `Err(MissingHeader("x-repo-commit"))` for the same test.

curling the URL with redirects `curl -v -L https://huggingface.co/TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf` confirms the same.

<details>
    <summary>curl -v -L https://huggingface.co/TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf</summary>
<pre>
*   Trying 2600:9000:263e:b000:17:b174:6d00:93a1:443...
* Connected to huggingface.co (2600:9000:263e:b000:17:b174:6d00:93a1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=huggingface.co
*  start date: Jun 13 00:00:00 2023 GMT
*  expire date: Jul 11 23:59:59 2024 GMT
*  subjectAltName: host "huggingface.co" matched cert's "huggingface.co"
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M01
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* Using Stream ID: 1 (easy handle 0x555a55796e90)
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> GET /TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf HTTP/2
> Host: huggingface.co
> user-agent: curl/7.81.0
> accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
< HTTP/2 302 
< content-type: text/plain; charset=utf-8
< content-length: 1151
< location: https://cdn-lfs.huggingface.co/repos/b0/ca/b0cae82fd4b3a362cab01d17953c45edac67d1c2dfb9fbb9e69c80c32dc2012e/08a5566d61d7cb6b420c3e4387a39e0078e1f2fe5f055f3a03887385304d4bfa?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27llama-2-7b-chat.Q4_K_M.gguf%3B+filename%3D%22llama-2-7b-chat.Q4_K_M.gguf%22%3B&Expires=1697819877&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY5NzgxOTg3N319LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy9iMC9jYS9iMGNhZTgyZmQ0YjNhMzYyY2FiMDFkMTc5NTNjNDVlZGFjNjdkMWMyZGZiOWZiYjllNjljODBjMzJkYzIwMTJlLzA4YTU1NjZkNjFkN2NiNmI0MjBjM2U0Mzg3YTM5ZTAwNzhlMWYyZmU1ZjA1NWYzYTAzODg3Mzg1MzA0ZDRiZmE%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qIn1dfQ__&Signature=bwDcDhinP6wilxOEC%7EPrOVVaG16FNpAv-L8zLw9RkcpoPu7wEZYJX3eIgcF1FHPdAipQH%7Ewvj8VGN0osS6AL9ECgt94hlfijxm9SXeXYTqUMc1Ru4kpCqBxm7iXr8hYWU8gcNIIPBMmsWW3ZnDQ0NzSvxz-oMfM95oK57A6R1%7EHsu3lXoYv7VjBSOJdF2Gblz5IVZdh7zlR2TgNqtKIEHIa9HZEn-l9xRJrs3yDMZ8S3ZMsRn-WA7%7E9UonPmptLKh8fvXUeYrQb6%7Ea8bY9kY6mh6DuD9WGr0TbjhnG6Xm9hFNHzDbyoQNCQOOyixwEbizRc2V796ont66LJvtlxc4A__&Key-Pair-Id=KVTP0A1DKRTAX
< date: Tue, 17 Oct 2023 17:07:12 GMT
< x-powered-by: huggingface-moon
< x-request-id: Root=1-652ebf40-06fc983d426f130c55ac4335
< access-control-allow-origin: https://huggingface.co
< vary: Origin, Accept
< access-control-expose-headers: X-Repo-Commit,X-Request-Id,X-Error-Code,X-Error-Message,ETag,Link,Accept-Ranges,Content-Range
< x-repo-commit: 191239b3e26b2882fb562ffccdd1cf0f65402adb
< accept-ranges: bytes
< x-linked-size: 4081004224
< x-linked-etag: "08a5566d61d7cb6b420c3e4387a39e0078e1f2fe5f055f3a03887385304d4bfa"
< x-cache: Miss from cloudfront
< via: 1.1 70bc8c513fe6574269b01e47be267a78.cloudfront.net (CloudFront)
< x-amz-cf-pop: SEA900-P3
< x-amz-cf-id: PGyO0d_U7N-Ln_vtsMLS6eRM2Frn1g28Qcx_Kx-0yHy1HQJMEoq31g==
< 
* Ignoring the response-body
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection #0 to host huggingface.co left intact
* Issue another request to this URL: 'https://cdn-lfs.huggingface.co/repos/b0/ca/b0cae82fd4b3a362cab01d17953c45edac67d1c2dfb9fbb9e69c80c32dc2012e/08a5566d61d7cb6b420c3e4387a39e0078e1f2fe5f055f3a03887385304d4bfa?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27llama-2-7b-chat.Q4_K_M.gguf%3B+filename%3D%22llama-2-7b-chat.Q4_K_M.gguf%22%3B&Expires=1697819877&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY5NzgxOTg3N319LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy9iMC9jYS9iMGNhZTgyZmQ0YjNhMzYyY2FiMDFkMTc5NTNjNDVlZGFjNjdkMWMyZGZiOWZiYjllNjljODBjMzJkYzIwMTJlLzA4YTU1NjZkNjFkN2NiNmI0MjBjM2U0Mzg3YTM5ZTAwNzhlMWYyZmU1ZjA1NWYzYTAzODg3Mzg1MzA0ZDRiZmE%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qIn1dfQ__&Signature=bwDcDhinP6wilxOEC%7EPrOVVaG16FNpAv-L8zLw9RkcpoPu7wEZYJX3eIgcF1FHPdAipQH%7Ewvj8VGN0osS6AL9ECgt94hlfijxm9SXeXYTqUMc1Ru4kpCqBxm7iXr8hYWU8gcNIIPBMmsWW3ZnDQ0NzSvxz-oMfM95oK57A6R1%7EHsu3lXoYv7VjBSOJdF2Gblz5IVZdh7zlR2TgNqtKIEHIa9HZEn-l9xRJrs3yDMZ8S3ZMsRn-WA7%7E9UonPmptLKh8fvXUeYrQb6%7Ea8bY9kY6mh6DuD9WGr0TbjhnG6Xm9hFNHzDbyoQNCQOOyixwEbizRc2V796ont66LJvtlxc4A__&Key-Pair-Id=KVTP0A1DKRTAX'
*   Trying 2600:9000:234d:c200:11:f807:5180:93a1:443...
* Connected to cdn-lfs.huggingface.co (2600:9000:234d:c200:11:f807:5180:93a1) port 443 (#1)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=cdn-lfs.huggingface.co
*  start date: Jul 27 00:00:00 2023 GMT
*  expire date: Aug 23 23:59:59 2024 GMT
*  subjectAltName: host "cdn-lfs.huggingface.co" matched cert's "cdn-lfs.huggingface.co"
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M01
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* Using Stream ID: 1 (easy handle 0x555a55796e90)
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> GET /repos/b0/ca/b0cae82fd4b3a362cab01d17953c45edac67d1c2dfb9fbb9e69c80c32dc2012e/08a5566d61d7cb6b420c3e4387a39e0078e1f2fe5f055f3a03887385304d4bfa?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27llama-2-7b-chat.Q4_K_M.gguf%3B+filename%3D%22llama-2-7b-chat.Q4_K_M.gguf%22%3B&Expires=1697819877&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY5NzgxOTg3N319LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5odWdnaW5nZmFjZS5jby9yZXBvcy9iMC9jYS9iMGNhZTgyZmQ0YjNhMzYyY2FiMDFkMTc5NTNjNDVlZGFjNjdkMWMyZGZiOWZiYjllNjljODBjMzJkYzIwMTJlLzA4YTU1NjZkNjFkN2NiNmI0MjBjM2U0Mzg3YTM5ZTAwNzhlMWYyZmU1ZjA1NWYzYTAzODg3Mzg1MzA0ZDRiZmE%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qIn1dfQ__&Signature=bwDcDhinP6wilxOEC%7EPrOVVaG16FNpAv-L8zLw9RkcpoPu7wEZYJX3eIgcF1FHPdAipQH%7Ewvj8VGN0osS6AL9ECgt94hlfijxm9SXeXYTqUMc1Ru4kpCqBxm7iXr8hYWU8gcNIIPBMmsWW3ZnDQ0NzSvxz-oMfM95oK57A6R1%7EHsu3lXoYv7VjBSOJdF2Gblz5IVZdh7zlR2TgNqtKIEHIa9HZEn-l9xRJrs3yDMZ8S3ZMsRn-WA7%7E9UonPmptLKh8fvXUeYrQb6%7Ea8bY9kY6mh6DuD9WGr0TbjhnG6Xm9hFNHzDbyoQNCQOOyixwEbizRc2V796ont66LJvtlxc4A__&Key-Pair-Id=KVTP0A1DKRTAX HTTP/2
> Host: cdn-lfs.huggingface.co
> user-agent: curl/7.81.0
> accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
< HTTP/2 200 
< content-type: binary/octet-stream
< content-length: 4081004224
< last-modified: Mon, 04 Sep 2023 16:42:07 GMT
< x-amz-storage-class: INTELLIGENT_TIERING
< x-amz-server-side-encryption: AES256
< x-amz-version-id: 0fUfUZjqztjlqlMos4m3EYw.QOX4675l
< content-disposition: attachment; filename*=UTF-8''llama-2-7b-chat.Q4_K_M.gguf; filename="llama-2-7b-chat.Q4_K_M.gguf";
< accept-ranges: bytes
< server: AmazonS3
< date: Tue, 17 Oct 2023 17:00:11 GMT
< etag: "b64bd51890def4d2f4becc5ee1bc62db-256"
< x-cache: Hit from cloudfront
< via: 1.1 5397b304713f6301c7c94ac084b6ed08.cloudfront.net (CloudFront)
< x-amz-cf-pop: SEA73-P1
< x-amz-cf-id: 1Tmo_3m7hY1xQ8EJro8yrK0myrUdfLFiHKiO5SvzGlZmSIiZ5K3zbQ==
< age: 33496
< cache-control: public, max-age=604800, immutable, s-maxage=604800
< vary: Origin
< 
Warning: Binary output can mess up your terminal. Use "--output -" to tell 
Warning: curl to output it to your terminal anyway, or consider "--output 
Warning: <FILE>" to save to a file.
* Failure writing output to destination
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* stopped the pause stream!
* Connection #1 to host cdn-lfs.huggingface.co left intact
</pre>
</details>

Note that `x-repo-commit` does get sent in one of the earlier redirects (and is included in `access-control-expose-headers`. It just does not appear in the final response.

I cannot find a way to build a request client to propagate headers on redirect - but at that point I'm a little out of my depth nor am sure if that is the correct solution.

I will submit a PR with the partial `redirect` / `etag` test + fix if it's wanted.

### PR DESCRIPTION
**Pull request by OlivierDehaene**
Close #33 

See internal: https://huggingface.slack.com/archives/C04U4U3UK24/p1700830641390659?thread_ts=1700489513.676129&cid=C04U4U3UK24

### CODE REVIEW

**Review by xtuc — APPROVED** [review]

File: src/api/sync.rs [hunk 1]
```diff
@@ -270,12 +264,51 @@ impl Api {
...
+                StatusCode::from_u16(status_code).unwrap(),
+                StatusCode::MOVED_PERMANENTLY
+                    | StatusCode::FOUND
+                    | StatusCode::SEE_OTHER
+                    | StatusCode::TEMPORARY_REDIRECT
+                    | StatusCode::PERMANENT_REDIRECT
+            )
+        };
+
+        // Follow redirects until host does not match if the first host
```
Comment (Wauplin) [inline src/api/sync.rs:288]:
```suggestion
        // Follow redirects until host does not match with the first host
```

File: src/api/sync.rs [hunk 2]
```diff
@@ -270,12 +264,51 @@ impl Api {
...
+                // Get header value
+                if let Some(location) = response.header("Location") {
+                    let url = Uri::from_str(location).unwrap();
+                    // Match hosts
+                    if url.host() == first_uri.host() {
+                        // Follow redirect
+                        response = self
+                            .no_redirect_client
+                            .get(response.get_url())
+                            .set(RANGE, "bytes=0-0")
```
Comment (Wauplin) [inline src/api/sync.rs:302]: do you happen to know why the GET range 0-0 instead of a HEAD?

**Review by Wauplin — COMMENTED** [review]
Conceptual approval :+1:

File: src/api/sync.rs (hunk 2 above)
Comment (Narsil) [reply src/api/sync.rs:302]: Because specific DNS routing rules with headers.
Although I'm not 100% sure if it's still needed.

**Review by Narsil — APPROVED** [review]
LGTM. 

I was under the impression we didn't need such rules anymore though.

**Comment by OlivierDehaene** [comment]
We still need these rules when we get a redirect because of casing issue.
I added a test for this specific problem

**Comment by jtsai-quid** [comment]
hi @OlivierDehaene, 
Is there an expected date for the next release?
Many thanks

**Comment by zackangelo** [comment]
Can you release this fix to crates.io? I'm unable to download some models using it because huggingface URLs are returning redirects in some cases.

**Comment by LaurentMazare** [comment]
+1, @Narsil (or others) if you could make a release to crates.io, that would be great! Fwiw we get some errors `Error: Header etag is missing` on the llama-3.1 repos.

**Pull request merged**